### PR TITLE
Add a root span to the client

### PR DESCRIPTION
**Pull request by poljar**
This allows us to have a nice single tree view of the whole client run instead of having traces scattered around:

image

Not all methods have been handled, and for some methods we would need to apply the root as a parent only if there isn't a previous parent, but this should be a good start.

Requires: https://github.com/matrix-org/matrix-rust-sdk/pull/1446

### CODE REVIEW

File: crates/matrix-sdk/src/client/login_builder.rs [hunk 1]
```diff
@@ -142,6 +142,7 @@ impl LoginBuilder {
     /// Instead of calling this function and `.await`ing its return value, you
     /// can also `.await` the `LoginBuilder` directly.
     #[instrument(
+        parent = &self.client.root_span,
```
Comment (jplatte) [inline crates/matrix-sdk/src/client/login_builder.rs:145]: We should do the same thing for `SsoLoginBuilder`, I think.

File: crates/matrix-sdk/src/client/mod.rs [hunk 2]
```diff
@@ -2147,7 +2174,7 @@ impl Client {
     ///     .await;
     /// })
     /// ```
-    #[instrument(skip(self, callback))]
+    #[instrument(skip_all, parent = &self.root_span)]
     pub async fn sync_with_callback<C>(
```
Comment (jplatte) [inline crates/matrix-sdk/src/client/mod.rs:2178]: This is a method that gets called from other sync methods. Using `parent =` here will make that relation invisible in logs, right?

**Comment by poljar** [comment]
One of the problems here is that our root span ~never gets uploaded. My comment that we can drop the span and get it emitted is false.

This seems to be related: https://github.com/open-telemetry/opentelemetry-rust/issues/888.

**Comment by poljar** [comment]
This PR proved that we want to have a root span, though I don't think a library should create root spans. What we instead want is to let EX create a root span and have that either be automatically propagated over the FFI or let the `ClientBuilder` take a root span.

Will happen in a different PR.

File: crates/matrix-sdk/src/client/mod.rs [hunk 3]
```diff
@@ -2244,11 +2271,15 @@ impl Client {
...
         loop {
+            debug!("Syncing");
             let result = self.sync_loop_helper(&mut sync_settings).await;
 
+            debug!("Running callback");
             if callback(result).await? == LoopCtrl::Break {
+                debug!("Callback told us to stop");
                 break;
             }
+            debug!("Done running callback");
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk/src/client/mod.rs]: Done in https://github.com/matrix-org/matrix-rust-sdk/pull/1447/commits/da28114792348d052c40aab75fa7821e2b2bf559.

File: crates/matrix-sdk/src/client/login_builder.rs (hunk 1 above)
Comment (poljar) [reply crates/matrix-sdk/src/client/login_builder.rs:145]: Yep, added in https://github.com/matrix-org/matrix-rust-sdk/pull/1447/commits/da28114792348d052c40aab75fa7821e2b2bf559.

File: crates/matrix-sdk/src/client/builder.rs [hunk 4]
```diff
@@ -326,8 +341,12 @@ impl ClientBuilder {
     ///   server discovery request is made which can fail; if you didn't set
     ///   [`server_versions(false)`][Self::server_versions], that amounts to
     ///   another request that can fail
+    #[instrument(skip_all, parent = &self.root_span,  target = "matrix_sdk::client", fields(homeserver))]
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk/src/client/builder.rs]: Fixed in https://github.com/matrix-org/matrix-rust-sdk/pull/1447/commits/da28114792348d052c40aab75fa7821e2b2bf559.

**Review by jplatte — APPROVED** [review]

**Comment by richvdh** [comment]
For archaeology:

> Will happen in a different PR.

I think the PR in question was #1692.

**Pull request merged**